### ISSUE TITLE
Allow to change the behavior for interfaces for rule IDE0040: Accessibility modifiers required

### ISSUE THREAD

**Issue by m-gasser** [dotnet/roslyn#77253]
**Brief description:**

In #74245 IDE0040 was fixed to also apply to interfaces. However it does so in a way that prohibits the configuration we actually want. We ideally would like this behavior:
1. Accessibility modifiers for members of non interface types are mandatory
2. Accessibility modifiers for nested types inside interface members are mandatory
3. Accessibility modifiers for members (methods, properties...) are not enforced but allowed.

This could be achieved by either changing the behavior of the for_non_interface_members option or by providing another option or making a separate rule for interfaces than for other types. I am unsure which option is the best way to go.

**Languages applicable:**

C# (and probably also VB)

**Code example that the analyzer should report:**

```
public interface IFoo
{
    int SomeMethod1(); // This is allowed as "customary right" because in older C# versions access modifiers could not be specified
    public int SomeMethod2(); // This is allowed for additional clarity or to be consistent with members of non interface types

    record Bar; // <- diagnostics applies here. For nested types, access modifiers should be applied no matter whether they are nested in an interface or another type
}
```

**Documentation requirements:**

When this analyzer is implemented, it must be documented by following the steps at [Documentation for IDE CodeStyle analyzers](https://github.com/dotnet/roslyn/blob/main/docs/contributing/Documentation%20for%20IDE%20CodeStyle%20analyzers.md).

**Comment by CyrusNajmabadi** [comment]
The first two items seem reasonable. The last is very odd. We don't generally have a rule that conditionally enforces in that manner. The idea with these rules is to ensure total consistency. The third rule would allow different parts of the codebase to differ with accessibility modifiers. We will discuss it, but we'd need a strong argument as to why it's desirable to enforce for everything else, but carve out this subset to not enforce.

**Comment by m-gasser** [comment]
Now that I thought more about it, yes it is a bit weird. Especially if there still would be code fixes in the IDE to add/remove the modifiers. It would probably be more reasonable to specify a preference for that case but with a separate severity. So we would configure the general rule to warning and the non-type interface members to information with a preference of no accessibility modifier.

**Comment by CyrusNajmabadi** [comment]
> Again I feel like the wrinkle on this rule which doesn't exist for most others is the fact that modifiers were not allowed in the past

Correct. That's why there are two options here. One to keep code looking like how it was written for like 20 years (`for_non_interface_members`). And one to be more

**Comment by manfred-brands** [comment]
I also just noticed this breaking code on my builds. 

We have the `for_non_interface_members` option enabled, but indeed some newer code uses access modifiers on interfaces.
However the rule now fires on interfaces with the error 'Accessibility modifiers required' even though they are there:

![Image](https://github.com/user-attachments/assets/c840ec12-09ed-439f-b4f9-8df8498812a1)

To me `dotnet_style_require_accessibility_modifiers = for_non_interface_members` means that accessibility modifiers are required everywhere, except on interfaces. It doesn't say they are not allowed on interface members.

**Comment by CyrusNajmabadi** [comment]
Covered here: https://github.com/dotnet/roslyn/blob/main/docs/compilers/CSharp/Compiler%20Breaking%20Changes%20-%20DotNet%209.md#dotnet_style_require_accessibility_modifiers-now-consistently-applies-to-interface-members

> means that accessibility modifiers are required everywhere, except on interfaces. 

The rules are intended to enforce consistent modifiers *everywhere*. The difference between 'always' and 'for_non_interface_members' is if pubic interface members should be required to can't the redundant 'public' modifier, or they should be disallowed from having it.

**Comment by CyrusNajmabadi** [comment]
Closing out as "are not enforced" is a non goal of our style options. 

This is especially the case when the primary goal is consistency. Having some interfaces use accessibility modifiers and some not leads to an inconsistent codebase. 

We do have a historical carve out for interfaces to not list the public modifier (given 20 years of .net where that was the only way to write them). But we do not want to have a carve out for interfaces being inconsistent among themselves.
